### DISCUSSION TITLE
[JS201] Closure 的 Scope 屬性及 scopeChain

### DISCUSSION THREAD

**Discussion by EshauHuang** [Lidemy/forum#165] [程式導師實驗計畫] [answered]
各位好～

在影片上瞭解到了宣告一個 function 時，會設定一個 [[Scope]] 的屬性

```
function test() {
...
}

test()

test.[[Scope]] = [globalEC.VO]
```
而當我們執行 `test`，就會設定 `testEC`，而這時會把 `test.[[Scope]]` 加進 `scopeChain` 內，使該 function 可以使用全域變數，這邊想問的第一個問題是，這個 `test.[[Scope]]` 會在什麼時候被刪除？是 `scopeChain` 取用後或執行完 `test` 後 ，還是整個程式碼執行完才刪除？

另一個問題是如果 `test` 內再宣告一個 function `inner`，而 `test` 執行完會回傳 `inner`，並在外面用一個變數接住後再執行 `inner`

```
function test() {
  function inner() {
    ...
  }
  return inner
}

var inner = test()
inner()
```

這邊在思考的點是 `testEC.AO` 存在的原因是 `inner.[[Scope]]` 還是 `innerEC.scopeChain`，而兩者差在後者多了 `innerEC.AO`，如果是後者的話，當 inner 宣告了一個很大的 obj，這樣是不是也會造成程式執行完後，沒有釋放記憶體的問題？

先感謝各位的解答！

**Comment by aszx87410** [comment, marked as answer]
> 是 scopeChain 取用後或執行完 test 後 ，還是整個程式碼執行完才刪除？

這邊我不確定，牽涉到底層怎麼實作，但我猜應該是整個程式碼執行完之後才會刪除

> 這邊在思考的點是 testEC.AO 存在的原因是 inner.[[Scope]] 還是 innerEC.scopeChain，而兩者差在後者多了 innerEC.AO，如果是後者的話，當 inner 宣告了一個很大的 obj，這樣是不是也會造成程式執行完後，沒有釋放記憶體的問題？

你說的兩個其實是一樣的，因為 inner.[[Scope]] = testEC.scopeChain = [testEC.AO, globalEC.VO]，然後對，inner EC 如果宣告了很大的 object，但在這整段程式碼執行完畢以前，記憶體不會被釋放

**Reply by EshauHuang** [reply]
謝謝老師，所以與 test 一樣，我們也不該在要回傳的 function 內宣告太大的變數。
